### ISSUE TITLE
Data file directory layout different between go_test and go_binary?

### ISSUE THREAD

**Issue by DeCarabas** [bazelbuild/rules_go#3217]
### What version of rules_go are you using?

0.31.0

### What version of gazelle are you using?

0.25.0

### What version of Bazel are you using?

5.1.0

### Does this issue reproduce with the latest releases of all the above?

Possibly.

### What operating system and processor architecture are you using?

MacOS Monterey, M1

### Any other potentially useful information about your toolchain?

Not for this

### What did you do?

I made a little go project, in a subdirectory of my workspace called `bobbin`. `bobbin/BUILD.bazel` contains:

```starlark
load("@io_bazel_rules_go//go:def.bzl", "go_binary", "go_library", "go_test")

go_library(
    name = "bobbin_lib",
    srcs = [
        "main.go",
        "root.go",
        "routes.go",
    ],
    data = [
        ":root.html",
    ],
    visibility = ["//visibility:private"],
    deps = [
        "@com_github_sirupsen_logrus//:logrus",
    ],
)

go_test(
    name = "bobbin_test",
    srcs = ["root_test.go"],
    embed = [":bobbin_lib"],
)

go_binary(
    name = "bobbin",
    embed = [":bobbin_lib"],
    visibility = ["//visibility:public"],
)
```

It implements a little HTTP server. Inside my handler, I have this:

```go
package main

import (
	"net/http"
)

func Root(w http.ResponseWriter, req *http.Request) {
	if req.URL.Path != "/" {
		http.NotFound(w, req)
		return
	}
	if req.Method != http.MethodGet {
		http.Error(w, "Method not supported", http.StatusMethodNotAllowed)
		return
	}
	http.ServeFile(w, req, "./bobbin/root.html")
}
```

The interesting and relevant part of that example is `http.ServeFile(w, req, "./bobbin/root.html")`; when I run my binary with `bazel run //bobbin` then everything works as expected, and my `root.html` file is in a directory named `bobbin` relative to my current working directory. (I have also packaged this binary into a container, via `go_image`, and the same directory structure is maintained.)

**However,** when I try to run this code via `bazel test //bobbin/...`, I find that `root.html` is *not* in a subdirectory named `bobbin`, and instead is in the same directory as my test binary. This causes that handler to get very confused.

Why is this the case? What's going on?

### What did you expect to see?

The directory layout of `data` files is the same between `go_binary` and `go_test`.

### What did you see instead?

They are different.

**Comment by achew22** [comment]
What you're experiencing is the difference between the way that go and bazel expect runfiles to work. Unfortunately they are not identical.

There is a flag on [`go_test`](https://github.com/bazelbuild/rules_go/blob/master/docs/go/core/rules.md#go_test) called `rundir` that affects this behavior. 

> A directory to cd to before the test is run. This should be a path relative to the execution dir of the test.The default behaviour is to change to the workspace relative path, this replicates the normal behaviour of go test so it is easy to write compatible tests.Setting it to . makes the test behave the normal way for a bazel test.Note: This defaults to the package path.

Please reopen if you think this is insufficient.

**Comment by DeCarabas** [comment]
No, this makes sense, thank you.

I might have chosen the other default, but now I know. Thanks!

On Jun 28, 2022, at 9:47 AM, Andrew Z Allen ***@***.***> wrote:

﻿
Closed #3217 as completed.

**Comment by achew22** [comment]
Unfortunately we can't "choose the other default" because then all of the tests that people have written before moving to Bazel (and from all of your 3rd party deps) will be broken. The default has to be the one that we picked. Sorry
